### PR TITLE
ddl: allow enum to append values

### LINKED ISSUE #6398: Support modify column to enum types.
TiDB :
```
CREATE TABLE ti1(a INT NOT NULL, b INT, c INT) engine=InnoDB;
ALTER TABLE ti1 ADD COLUMN e ENUM('a', 'b') FIRST;
ALTER TABLE ti1 MODIFY COLUMN e ENUM('a', 'b', 'c');
```

reports `Error 1105: unsupported modify column modify enum column is not supported`

### PR DESCRIPTION
**Pull request by mail2fish**
## What have you changed? (mandatory)

fix #6398

Allow to append new values to existing ENUM fields.
Removing values or changing existing values is still unsupported.

## What are the type of the changes (mandatory)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (mandatory)?

- added unit test cases in ddl/db_change_test.go

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1430,6 +1430,7 @@ func modifiable(origin *types.FieldType, to *types.FieldType) error {
 	if to.Charset != origin.Charset {
 		msg := fmt.Sprintf("charset %s not match origin %s", to.Charset, origin.Charset)
 		return errUnsupportedModifyColumn.GenWithStackByArgs(msg)
+
```
Comment (crazycs520) [inline ddl/ddl_api.go]: remove this line.

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -1455,7 +1456,22 @@ func modifiable(origin *types.FieldType, to *types.FieldType) error {
...
-		return errUnsupportedModifyColumn.GenWithStackByArgs("modify enum column is not supported")
+		if origin.Tp == to.Tp {
+			if len(to.Elems) < len(origin.Elems) {
+				msg := fmt.Sprintf("elem length %d is less than origin %d", len(to.Elems), len(origin.Elems))
+				return errUnsupportedModifyColumn.GenWithStackByArgs(msg)
+			}
+			for index, originElem := range origin.Elems {
+				toElem := to.Elems[index]
+				if originElem != toElem {
+					msg := fmt.Sprintf("cannot modify value %s to %s", originElem, toElem)
```
Comment (crazycs520) [inline ddl/ddl_api.go]: How about:
msg := fmt.Sprintf("cannot modify enum column value %s to %s", originElem, toElem)

File: ddl/ddl_api.go [hunk 3]
```diff
@@ -1455,7 +1456,22 @@ func modifiable(origin *types.FieldType, to *types.FieldType) error {
 			return nil
 		}
 	case mysql.TypeEnum:
-		return errUnsupportedModifyColumn.GenWithStackByArgs("modify enum column is not supported")
+		if origin.Tp == to.Tp {
+			if len(to.Elems) < len(origin.Elems) {
+				msg := fmt.Sprintf("elem length %d is less than origin %d", len(to.Elems), len(origin.Elems))
```
Comment (crazycs520) [inline ddl/ddl_api.go]: How about:
msg := fmt.Sprintf("enum column elements length %d is less than origin %d", len(to.Elems), len(origin.Elems))

File: ddl/db_change_test.go [hunk 4]
```diff
@@ -349,6 +349,60 @@ type expectQuery struct {
...
+	checkResult(result, testkit.Rows(expected...))
+
+	_, err = s.se.Execute(context.Background(), "update t set c2='N' where c4 = 10")
+	c.Assert(err, IsNil)
+
+	result, err = s.execQuery(tk, "select c2 from t where c4 = 10")
+	c.Assert(err, IsNil)
+
+	expected = []string{"8 N", "10 N", "11 A"}
+
```
Comment (mccxj) [inline ddl/db_change_test.go]: so many blank lines

**Comment by mail2fish** [comment]
@winoros This PR is needed by my job. Maybe I could post another PR for the set.

File: ddl/ddl_api.go [hunk 5]
```diff
@@ -1455,7 +1455,22 @@ func modifiable(origin *types.FieldType, to *types.FieldType) error {
 			return nil
 		}
 	case mysql.TypeEnum:
-		return errUnsupportedModifyColumn.GenWithStackByArgs("modify enum column is not supported")
+		if origin.Tp == to.Tp {
+			if len(to.Elems) < len(origin.Elems) {
+				msg := fmt.Sprintf("enum column elements length %d is less than origin %d", len(to.Elems), len(origin.Elems))
```
Comment (winoros) [inline ddl/ddl_api.go]: `The number of enum column's elements is less than the original`.

File: ddl/ddl_api.go [hunk 6]
```diff
@@ -1455,7 +1455,22 @@ func modifiable(origin *types.FieldType, to *types.FieldType) error {
...
+				return errUnsupportedModifyColumn.GenWithStackByArgs(msg)
+			}
+			for index, originElem := range origin.Elems {
+				toElem := to.Elems[index]
+				if originElem != toElem {
+					msg := fmt.Sprintf("cannot modify enum column value %s to %s", originElem, toElem)
```
Comment (winoros) [inline ddl/ddl_api.go]: `cannot modify enum column's element %s to %s`.

File: ddl/ddl_api.go [hunk 7]
```diff
@@ -1455,7 +1455,22 @@ func modifiable(origin *types.FieldType, to *types.FieldType) error {
...
+					return errUnsupportedModifyColumn.GenWithStackByArgs(msg)
+				}
+			}
+			return nil
+		}
+		msg := fmt.Sprintf("cannot modify enum type column to type %s", to.String())
```
Comment (winoros) [inline ddl/ddl_api.go]: `cannot modify enum column's type to type %s`

File: ddl/db_change_test.go [hunk 8]
```diff
@@ -349,6 +349,49 @@ type expectQuery struct {
 	rows []string
 }
 
+func (s *testStateChangeSuite) TestAppendEnum(c *C) {
```
Comment (zz-jason) [inline ddl/db_change_test.go]: I think we should test the `ALTER TABLE ... MODIFY COLUMN` statements which modify the enum column to another column type.

Comment (mail2fish) [reply ddl/db_change_test.go]: What's the test result you expected? An error message?

Comment (zz-jason) [reply ddl/db_change_test.go]: According to the `modifiable()` function, some changes are allowed, while others are not.

Comment (mail2fish) [reply ddl/db_change_test.go]: At this time, to.Tp is always be enum.  If we alter enum's column to other type, we will got the error message:  the number of enum column's elements is less than the original. 

Is that all right?

Comment (zz-jason) [reply ddl/db_change_test.go]: Yes, for example:
```sql
CREATE TABLE ti1(a INT NOT NULL, b INT, c INT, e ENUM('a', 'b')) engine=InnoDB;

ALTER TABLE ti1 MODIFY COLUMN e ENUM('a', 'b', 'c'); -- succeed 
ALTER TABLE ti1 MODIFY COLUMN e ENUM('a'); -- failed
```

Comment (mail2fish) [reply ddl/db_change_test.go]: Hi, I have add some new test, please check it. thank you.

File: ddl/ddl_api.go [hunk 9]
```diff
@@ -1455,7 +1455,22 @@ func modifiable(origin *types.FieldType, to *types.FieldType) error {
...
 	case mysql.TypeEnum:
-		return errUnsupportedModifyColumn.GenWithStackByArgs("modify enum column is not supported")
+		if origin.Tp == to.Tp {
+			if len(to.Elems) < len(origin.Elems) {
+				msg := fmt.Sprintf("the number of enum column's elements is less than the original: %d", len(origin.Elems))
+				return errUnsupportedModifyColumn.GenWithStackByArgs(msg)
+			}
+			for index, originElem := range origin.Elems {
+				toElem := to.Elems[index]
+				if originElem != toElem {
```
Comment (imtbkcat) [inline ddl/ddl_api.go]: For enum `enum('N','Y')`, `alter table t change c2 c2 enum('N','Y','A','B') DEFAULT 'A';` is okay, but `alter table t change c2 c2 enum('A','B','N','Y') DEFAULT 'A'` does not work.

Comment (mail2fish) [reply ddl/ddl_api.go]: What do you mean? The enum order shouldn't be changed. The SQL "alter table t change c2 c2 enum('A','B','N','Y') DEFAULT 'A' " is also  not working at MySQL.

Comment (imtbkcat) [reply ddl/ddl_api.go]: I use MySQL 5.6 to execute these case.
in MySQL:
   
    mysql> create table t (c1 varchar(64),c2 enum('N','Y') not null default 'N',c3 
    timestamp on update current_timestamp,c4 int primary key,unique key idx2 (c2, 
    c3));
    Query OK, 0 rows affected (0.03 sec)
    mysql> insert into t values('a', 'N', '2017-07-01', 8);
    Query OK, 1 row affected (0.00 sec)
    mysql> alter table t change c2 c2 enum('A','B','N','Y') DEFAULT 'A';
    Query OK, 1 row affected (0.05 sec)
    Records: 1  Duplicates: 0  Warnings: 0
in TiDB:  

    mysql> insert into t values('a', 'N', '2017-07-01', 8);
    Query OK, 1 row affected (0.00 sec)
    mysql> alter table t change c2 c2 enum('A','B','N','Y') DEFAULT 'A';
    ERROR 1105 (HY000): unsupported modify column cannot modify enum column 
    value N to A
    mysql> alter table t change c2 c2 enum('N','Y','A','B') DEFAULT 'A';
    Query OK, 0 rows affected (0.01 sec)

**Comment by zz-jason** [comment]
@zimulala @winkyao PTAL

**Comment by mail2fish** [comment]
You are right. It worked in MySQL.

But I am not sure whether we should support this feature.

After I changed the enum order in MySQL 5.6, The enum's value in the rows
has changed.


mysql> create table t (c1 varchar(64),c2 enum('N','Y') not null default
'N',c3  timestamp on update current_timestamp,c4 int primary key,unique key
idx2 (c2,  c3));
Query OK, 0 rows affected (0.21 sec)

mysql> insert into t values('a', 'N', '2017-07-01', 8);
Query OK, 1 row affected (0.04 sec)

mysql> select c2+0 from t;
+------+
| c2+0 |
+------+
|    1 |
+------+
1 row in set (0.00 sec)

mysql> alter table t change c2 c2 enum('A','B','D','F','Y','N') DEFAULT 'D';
Query OK, 1 row affected (0.53 sec)
Records: 1  Duplicates: 0  Warnings: 0

mysql> select c2+0 from t;
+------+
| c2+0 |
+------+
|    6 |
+------+
1 row in set (0.00 sec)


Lingyu Song <notifications@github.com> 于2018年9月26日周三 下午4:59写道：

**Comment by coocood** [comment]
LGTM

**Pull request merged**